c/misra/src/rules/RULE-21-25/InvalidMemoryOrderArgument.ql

```diff
@@ -0,0 +1,172 @@
+          extraParams >= 0 and
+          orderParamIdx = baseParamIdx + extraParams
+        )
+        or
+        // Clang case, no inserted parameters:
+        // __c11_atomic_load(object, order)
+        suffix = "" and
+        prefix = "__c11_" and
+        orderParamIdx = baseParamIdx
+        or
+        // Non-macro case, may occur in a subset of gcc/clang functions:
+        prefix = "" and
+        suffix = "_explicit" and
+        orderParamIdx = baseParamIdx
+      )
+    )
+  }
+
+  int getOrderParameterIdx() { result = orderParamIdx }
+}
+
+module MemoryOrderFlowConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node node) {
+    // Direct usage of memory order constant
+    exists(MemoryOrderConstantAccess constant |
+      node.asExpr() = constant and
+      not constant.isAllowedOrder()
+    )
+    or
+    // A literal with a disallowed constant integer value
+    exists(Literal literal |
+      node.asExpr() = literal and
+      not literal.getValue().toInt() = any(AllowedMemoryOrder a).getValue().toInt()
+    )
+    or
+    // Everything else: not a memory order constant or an integer valued literal, also exclude
+    // variables and functions, things that flow further back.
+    exists(Expr e |
+      node.asExpr() = e and
+      not e instanceof MemoryOrderConstantAccess and
+      not e instanceof Literal and
+      not e instanceof VariableAccess and
+      not e instanceof FunctionCall and
+      not DataFlow::localFlowStep(_, node)
+    )
+  }
+
+  predicate isSink(DataFlow::Node node) {
+    exists(FunctionCall fc |
+      node.asExpr() =
+        fc.getArgument(fc.getTarget().(MemoryOrderedStdAtomicFunction).getOrderParameterIdx())
+    )
+  }
+}
+
+module MemoryOrderFlow = DataFlow::Global<MemoryOrderFlowConfig>;
+
+import MemoryOrderFlow::PathGraph
+
+/**
+ * If the node is a memory order constant, or shares a value with a memory order constant, then
+ * return the name of that cnonstant. Otherwise, simply print the node.
+ */
+string describeMemoryOrderNode(DataFlow::Node node) {
+  if node.asExpr() instanceof MemoryOrderConstantExpr
+  then result = node.asExpr().(MemoryOrderConstantExpr).getMemoryOrderString()
+  else result = node.toString()
+}
+
+from
+  Expr argument, Function function, string value, MemoryOrderFlow::PathNode source,
+  MemoryOrderFlow::PathNode sink
+where
+  not isExcluded(argument, Concurrency6Package::invalidMemoryOrderArgumentQuery()) and
+  MemoryOrderFlow::flowPath(source, sink) and
+  argument = sink.getNode().asExpr() and
+  value = describeMemoryOrderNode(source.getNode()) and
+  // Double check that we didn't find flow from something equivalent to the allowed value.
+  not value = any(AllowedMemoryOrder e).getName() and
+  function.getACallToThisFunction().getAnArgument() = argument
+select argument, source, sink, "Invalid memory order '$@' in call to function '$@'.", value, value,
+  function, function.toString()
```

Comment:
In general prefer to use `.getName()` for functions, as `toString()` may be evaluated for all elements in the database.
```suggestion
  function, function.getName()
```